{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initial processing to prepare the sequences for ESM-2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df  = pd.read_csv(\"../data/intermediate/variant_protein_sequences.txt\", sep=\"\\t\", dtype=\"str\")\n",
    "df = df.dropna(subset=[\"wt_seq\", \"ref_aa\", \"alt_aa\", \"pos_aa\", \"mut_seq\"], how=\"any\").drop_duplicates()\n",
    "\n",
    "# Remove the stop codons\n",
    "df[\"wt_seq\"] = df[\"wt_seq\"].str.rstrip(\"*\")\n",
    "df[\"mut_seq\"] = df[\"mut_seq\"].str.rstrip(\"*\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Update the sequences to only keep the +/- 127 residues from the pos_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"pos_aa\"] = pd.to_numeric(df[\"pos_aa\"], errors=\"coerce\").astype(\"Int64\")\n",
    "\n",
    "flank = 127\n",
    "\n",
    "def extract_subsequence(full_seq, pos, flank=flank):\n",
    "    if pd.isna(full_seq) or pos is None:\n",
    "        return None\n",
    "    \n",
    "    pos_idx = pos - 1 \n",
    "    start = max(0, pos_idx - flank)\n",
    "    end = min(len(full_seq), pos_idx + flank + 1) \n",
    "    \n",
    "    return full_seq[start:end]\n",
    "\n",
    "df[\"wt_seq\"] = df.apply(lambda row: extract_subsequence(row[\"wt_seq\"], row[\"pos_aa\"]), axis=1)\n",
    "df[\"mut_seq\"] = df.apply(lambda row: extract_subsequence(row[\"mut_seq\"], row[\"pos_aa\"]), axis=1)\n",
    "\n",
    "df.drop_duplicates(inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generate wild-type and mutant protein IDs for all variants in the variant set (these IDs are used for merging later)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_ID</th>\n",
       "      <th>ensg</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>pos_aa</th>\n",
       "      <th>wt_ID</th>\n",
       "      <th>mut_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10-100057090-C-T</td>\n",
       "      <td>ENSG00000120054</td>\n",
       "      <td>D</td>\n",
       "      <td>N</td>\n",
       "      <td>312</td>\n",
       "      <td>ENSG00000120054-312-D-wt</td>\n",
       "      <td>ENSG00000120054-312-D-N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10-100069757-C-T</td>\n",
       "      <td>ENSG00000120054</td>\n",
       "      <td>G</td>\n",
       "      <td>D</td>\n",
       "      <td>178</td>\n",
       "      <td>ENSG00000120054-178-G-wt</td>\n",
       "      <td>ENSG00000120054-178-G-D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10-100076062-G-A</td>\n",
       "      <td>ENSG00000120054</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>90</td>\n",
       "      <td>ENSG00000120054-90-A-wt</td>\n",
       "      <td>ENSG00000120054-90-A-V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10-100081405-G-A</td>\n",
       "      <td>ENSG00000120054</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>74</td>\n",
       "      <td>ENSG00000120054-74-P-wt</td>\n",
       "      <td>ENSG00000120054-74-P-L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10-100152307-T-C</td>\n",
       "      <td>ENSG00000107566</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>291</td>\n",
       "      <td>ENSG00000107566-291-I-wt</td>\n",
       "      <td>ENSG00000107566-291-I-V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408702</th>\n",
       "      <td>19-48318735-G-A</td>\n",
       "      <td>ENSG00000105479</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>50</td>\n",
       "      <td>ENSG00000105479-50-H-wt</td>\n",
       "      <td>ENSG00000105479-50-H-Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408703</th>\n",
       "      <td>19-48318737-A-G</td>\n",
       "      <td>ENSG00000105479</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>49</td>\n",
       "      <td>ENSG00000105479-49-V-wt</td>\n",
       "      <td>ENSG00000105479-49-V-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408704</th>\n",
       "      <td>19-48318744-T-G</td>\n",
       "      <td>ENSG00000105479</td>\n",
       "      <td>K</td>\n",
       "      <td>Q</td>\n",
       "      <td>47</td>\n",
       "      <td>ENSG00000105479-47-K-wt</td>\n",
       "      <td>ENSG00000105479-47-K-Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408705</th>\n",
       "      <td>19-48318746-C-G</td>\n",
       "      <td>ENSG00000105479</td>\n",
       "      <td>S</td>\n",
       "      <td>T</td>\n",
       "      <td>46</td>\n",
       "      <td>ENSG00000105479-46-S-wt</td>\n",
       "      <td>ENSG00000105479-46-S-T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408706</th>\n",
       "      <td>19-48318749-T-C</td>\n",
       "      <td>ENSG00000105479</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>45</td>\n",
       "      <td>ENSG00000105479-45-Y-wt</td>\n",
       "      <td>ENSG00000105479-45-Y-C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>408707 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              variant_ID             ensg ref_aa alt_aa  pos_aa  \\\n",
       "0       10-100057090-C-T  ENSG00000120054      D      N     312   \n",
       "1       10-100069757-C-T  ENSG00000120054      G      D     178   \n",
       "2       10-100076062-G-A  ENSG00000120054      A      V      90   \n",
       "3       10-100081405-G-A  ENSG00000120054      P      L      74   \n",
       "4       10-100152307-T-C  ENSG00000107566      I      V     291   \n",
       "...                  ...              ...    ...    ...     ...   \n",
       "408702   19-48318735-G-A  ENSG00000105479      H      Y      50   \n",
       "408703   19-48318737-A-G  ENSG00000105479      V      A      49   \n",
       "408704   19-48318744-T-G  ENSG00000105479      K      Q      47   \n",
       "408705   19-48318746-C-G  ENSG00000105479      S      T      46   \n",
       "408706   19-48318749-T-C  ENSG00000105479      Y      C      45   \n",
       "\n",
       "                           wt_ID                   mut_ID  \n",
       "0       ENSG00000120054-312-D-wt  ENSG00000120054-312-D-N  \n",
       "1       ENSG00000120054-178-G-wt  ENSG00000120054-178-G-D  \n",
       "2        ENSG00000120054-90-A-wt   ENSG00000120054-90-A-V  \n",
       "3        ENSG00000120054-74-P-wt   ENSG00000120054-74-P-L  \n",
       "4       ENSG00000107566-291-I-wt  ENSG00000107566-291-I-V  \n",
       "...                          ...                      ...  \n",
       "408702   ENSG00000105479-50-H-wt   ENSG00000105479-50-H-Y  \n",
       "408703   ENSG00000105479-49-V-wt   ENSG00000105479-49-V-A  \n",
       "408704   ENSG00000105479-47-K-wt   ENSG00000105479-47-K-Q  \n",
       "408705   ENSG00000105479-46-S-wt   ENSG00000105479-46-S-T  \n",
       "408706   ENSG00000105479-45-Y-wt   ENSG00000105479-45-Y-C  \n",
       "\n",
       "[408707 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_set = df[[\"ID\", \"ensg\", \"ref_aa\", \"alt_aa\", \"pos_aa\"]].copy()\n",
    "variant_set.rename(columns={\"ID\": \"variant_ID\"}, inplace=True)\n",
    "variant_set[\"wt_aa\"] = \"wt\"\n",
    "variant_set[\"wt_ID\"] = variant_set[[\"ensg\", \"pos_aa\", \"ref_aa\", \"wt_aa\"]].astype(str).agg(\"-\".join, axis=1)\n",
    "variant_set[\"mut_ID\"] = variant_set[[\"ensg\", \"pos_aa\", \"ref_aa\", \"alt_aa\"]].astype(str).agg(\"-\".join, axis=1)\n",
    "variant_set.drop(columns=[\"wt_aa\"], inplace=True)\n",
    "\n",
    "variant_set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Construct a df with all sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000120054-312-D-wt</td>\n",
       "      <td>LPDNWKSQVEPETRAVIRWMHSFNFVLSANLHGGAVVANYPYDKSF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000120054-178-G-wt</td>\n",
       "      <td>IGRSVEGRHLYVLEFSDHPGIHEPLEPEVKYVGNMHGNEALGRELM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000120054-90-A-wt</td>\n",
       "      <td>MSDLLSVFLHLLLLFKLVAPVTFRHHRYDDLVRTLYKVQNECPGIT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000120054-74-P-wt</td>\n",
       "      <td>MSDLLSVFLHLLLLFKLVAPVTFRHHRYDDLVRTLYKVQNECPGIT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000107566-291-I-wt</td>\n",
       "      <td>GLTIQAVRVTKPKIPEAIRRNFELMEAEKTKLLIAAQKQKVVEKEA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691451</th>\n",
       "      <td>ENSG00000105479-50-H-Y</td>\n",
       "      <td>MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRAYS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691452</th>\n",
       "      <td>ENSG00000105479-49-V-A</td>\n",
       "      <td>MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRAYS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691453</th>\n",
       "      <td>ENSG00000105479-47-K-Q</td>\n",
       "      <td>MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRAYS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691454</th>\n",
       "      <td>ENSG00000105479-46-S-T</td>\n",
       "      <td>MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRAYT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691455</th>\n",
       "      <td>ENSG00000105479-45-Y-C</td>\n",
       "      <td>MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRACS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>691456 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              ID  \\\n",
       "0       ENSG00000120054-312-D-wt   \n",
       "1       ENSG00000120054-178-G-wt   \n",
       "2        ENSG00000120054-90-A-wt   \n",
       "3        ENSG00000120054-74-P-wt   \n",
       "4       ENSG00000107566-291-I-wt   \n",
       "...                          ...   \n",
       "691451    ENSG00000105479-50-H-Y   \n",
       "691452    ENSG00000105479-49-V-A   \n",
       "691453    ENSG00000105479-47-K-Q   \n",
       "691454    ENSG00000105479-46-S-T   \n",
       "691455    ENSG00000105479-45-Y-C   \n",
       "\n",
       "                                                      seq  \n",
       "0       LPDNWKSQVEPETRAVIRWMHSFNFVLSANLHGGAVVANYPYDKSF...  \n",
       "1       IGRSVEGRHLYVLEFSDHPGIHEPLEPEVKYVGNMHGNEALGRELM...  \n",
       "2       MSDLLSVFLHLLLLFKLVAPVTFRHHRYDDLVRTLYKVQNECPGIT...  \n",
       "3       MSDLLSVFLHLLLLFKLVAPVTFRHHRYDDLVRTLYKVQNECPGIT...  \n",
       "4       GLTIQAVRVTKPKIPEAIRRNFELMEAEKTKLLIAAQKQKVVEKEA...  \n",
       "...                                                   ...  \n",
       "691451  MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRAYS...  \n",
       "691452  MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRAYS...  \n",
       "691453  MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRAYS...  \n",
       "691454  MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRAYT...  \n",
       "691455  MPLGRLAGSARSEEGSEAFLEGMVDWELSRLQRQCKVMEGERRACS...  \n",
       "\n",
       "[691456 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.drop(columns=[\"ID\"], inplace=True)\n",
    "wt = df.copy().drop(columns=[\"mut_seq\", \"alt_aa\"]).drop_duplicates(subset=[\"ensg\", \"ref_aa\", \"pos_aa\"], keep=\"first\")\n",
    "mut = df.copy().drop(columns=[\"wt_seq\"])\n",
    "mut[\"ID\"] = mut[[\"ensg\", \"pos_aa\", \"ref_aa\", \"alt_aa\"]].astype(str).agg(\"-\".join, axis=1)\n",
    "mut = mut[[\"ID\"] + [col for col in mut.columns if col != \"ID\"]]\n",
    "mut = mut.rename(columns={\"mut_seq\": \"seq\"})\n",
    "wt[\"alt_aa\"] = \"wt\"\n",
    "wt[\"ID\"] = wt[[\"ensg\", \"pos_aa\", \"ref_aa\", \"alt_aa\"]].astype(str).agg(\"-\".join, axis=1)\n",
    "wt = wt[[\"ID\"] + [col for col in wt.columns if col != \"ID\"]]\n",
    "wt = wt.rename(columns={\"wt_seq\": \"seq\"})\n",
    "wt = wt[mut.columns]\n",
    "sequences = pd.concat([wt, mut], ignore_index=True)[[\"ID\", \"seq\"]]\n",
    "del df\n",
    "\n",
    "sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute ESM-2 embeddings and perplexity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ozcelik2/miniconda3/envs/ml_env/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 691456/691456 sequences (100.00%)\n",
      "All sequences processed and saved to ../data/intermediate/esm2_output.pkl\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import torch\n",
    "import torch.nn.functional as F\n",
    "from transformers import EsmForMaskedLM, EsmTokenizer\n",
    "import math\n",
    "\n",
    "OUTPUT_PATH = \"../data/intermediate/esm2_output.pkl\"\n",
    "MODEL_NAME = \"facebook/esm2_t33_650M_UR50D\"\n",
    "\n",
    "tokenizer = EsmTokenizer.from_pretrained(MODEL_NAME)\n",
    "model = EsmForMaskedLM.from_pretrained(MODEL_NAME)\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "model.to(device)\n",
    "model.eval()\n",
    "\n",
    "def compute_output(seq):\n",
    "    \"\"\"\n",
    "    Compute mean-pooled protein embedding and perplexity for a given sequence.\n",
    "    \"\"\"\n",
    "    spaced_seq = \" \".join(seq)\n",
    "    inputs = tokenizer(spaced_seq, return_tensors=\"pt\", add_special_tokens=True, truncation=True)\n",
    "    inputs = {k: v.to(device) for k, v in inputs.items()}\n",
    "\n",
    "    with torch.no_grad():\n",
    "        outputs = model(**inputs, output_hidden_states=True)\n",
    "\n",
    "    # Mean-pooled embedding\n",
    "    last_hidden_state = outputs.hidden_states[-1]\n",
    "    protein_emb = last_hidden_state.mean(dim=1).squeeze(0).cpu().numpy()\n",
    "\n",
    "    # Perplexity\n",
    "    logits = outputs.logits\n",
    "    log_probs = F.log_softmax(logits, dim=-1)\n",
    "    input_ids = inputs[\"input_ids\"].squeeze(0)\n",
    "    true_token_log_probs = log_probs[0, torch.arange(input_ids.shape[0]), input_ids]\n",
    "    valid_log_probs = true_token_log_probs[1:-1]\n",
    "    mean_log_likelihood = valid_log_probs.mean().item()\n",
    "    perplexity = math.exp(-mean_log_likelihood)\n",
    "\n",
    "    return protein_emb, mean_log_likelihood, perplexity\n",
    "\n",
    "protein_embeddings = []\n",
    "mll_values = []\n",
    "ppl_values = []\n",
    "\n",
    "for idx, row in enumerate(sequences.itertuples(), start=1):\n",
    "    emb, mll, ppl = compute_output(row.seq)\n",
    "    protein_embeddings.append(emb)\n",
    "    mll_values.append(mll)\n",
    "    ppl_values.append(ppl)\n",
    "    print(f\"Processed {idx}/{len(sequences)} sequences ({idx/len(sequences)*100:.2f}%)\", end=\"\\r\")\n",
    "\n",
    "sequences[\"esm2_protein_embedding\"] = protein_embeddings\n",
    "sequences[\"esm2_mean_log_likelihood\"] = mll_values\n",
    "sequences[\"esm2_perplexity\"] = ppl_values\n",
    "\n",
    "os.makedirs(os.path.dirname(OUTPUT_PATH), exist_ok=True)\n",
    "sequences.to_pickle(OUTPUT_PATH)\n",
    "print(f\"\\nAll sequences processed and saved to {OUTPUT_PATH}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge ESM-2 embeddings and perplexity metrics for wild-type and mutant sequences\n",
    "\n",
    "This step merges ESM-2 outputs (embeddings, mean log-likelihood, and perplexity) from a shared dataframe (sequences) into the variant_set. Each variant has a wild-type (wt_ID) and mutant (mut_ID) identifier, which are used to join the corresponding metrics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "esm2_data = sequences[[\"ID\", \"esm2_protein_embedding\", \"esm2_mean_log_likelihood\", \"esm2_perplexity\"]].drop_duplicates(subset=[\"ID\"])\n",
    "\n",
    "# Merge ESM-2 metrics for wild-type sequences\n",
    "variant_set = variant_set.merge(\n",
    "    esm2_data,\n",
    "    left_on=\"wt_ID\",\n",
    "    right_on=\"ID\",\n",
    "    how=\"left\"\n",
    ").rename(columns={\n",
    "    \"esm2_protein_embedding\": \"wt_protein_embedding\",\n",
    "    \"esm2_mean_log_likelihood\": \"wt_esm2_mean_log_likelihood\",\n",
    "    \"esm2_perplexity\": \"wt_esm2_perplexity\"\n",
    "})\n",
    "variant_set.drop(columns=[\"ID\"], inplace=True)\n",
    "\n",
    "# Merge ESM-2 metrics for mutant sequences\n",
    "variant_set = variant_set.merge(\n",
    "    esm2_data,\n",
    "    left_on=\"mut_ID\",\n",
    "    right_on=\"ID\",\n",
    "    how=\"left\"\n",
    ").rename(columns={\n",
    "    \"esm2_protein_embedding\": \"mut_protein_embedding\",\n",
    "    \"esm2_mean_log_likelihood\": \"mut_esm2_mean_log_likelihood\",\n",
    "    \"esm2_perplexity\": \"mut_esm2_perplexity\"\n",
    "})\n",
    "variant_set.drop(columns=[\"ID\"], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute differences in perplexity and mean log-likelihood (mutant - wild-type)\n",
    "\n",
    "This section calculates the change in ESM-2 metrics between the mutant and wild-type protein sequences. These deltas can help assess how mutations impact sequence likelihood."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_set[\"esm2_mean_log_likelihood_difference\"] = (\n",
    "    variant_set[\"mut_esm2_mean_log_likelihood\"] - variant_set[\"wt_esm2_mean_log_likelihood\"]\n",
    ")\n",
    "\n",
    "variant_set[\"esm2_perplexity_difference\"] = (\n",
    "    variant_set[\"mut_esm2_perplexity\"] - variant_set[\"wt_esm2_perplexity\"]\n",
    ")\n",
    "\n",
    "variant_set.drop(columns=[\n",
    "    \"wt_esm2_mean_log_likelihood\", \"wt_esm2_perplexity\",\n",
    "    \"mut_esm2_mean_log_likelihood\", \"mut_esm2_perplexity\"\n",
    "], inplace=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute cosine similarity and average absolute difference between embeddings\n",
    "\n",
    "This step compares the high-dimensional protein embeddings from ESM-2 between the wild-type and mutant sequences. Cosine similarity captures directional changes in embedding space, while the average absolute difference reflects the magnitude of overall change."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import numpy as np\n",
    "\n",
    "def compute_cosine_similarity(wt_embedding, mut_embedding):\n",
    "    if wt_embedding is None or mut_embedding is None:\n",
    "        return np.nan\n",
    "    wt_embedding = np.array(wt_embedding).reshape(1, -1)\n",
    "    mut_embedding = np.array(mut_embedding).reshape(1, -1)\n",
    "    return cosine_similarity(wt_embedding, mut_embedding)[0][0]\n",
    "\n",
    "def compute_average_absolute_difference(wt_embedding, mut_embedding):\n",
    "    if wt_embedding is None or mut_embedding is None:\n",
    "        return np.nan\n",
    "    wt_embedding = np.array(wt_embedding)\n",
    "    mut_embedding = np.array(mut_embedding)\n",
    "    return np.mean(np.abs(wt_embedding - mut_embedding))\n",
    "\n",
    "variant_set[\"esm2_cosine_similarity\"] = variant_set.apply(\n",
    "    lambda row: compute_cosine_similarity(row[\"wt_protein_embedding\"], row[\"mut_protein_embedding\"]),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "variant_set[\"esm2_avg_abs_difference\"] = variant_set.apply(\n",
    "    lambda row: compute_average_absolute_difference(row[\"wt_protein_embedding\"], row[\"mut_protein_embedding\"]),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "variant_set.drop(columns=[\"wt_protein_embedding\", \"mut_protein_embedding\"], inplace=True)\n",
    "variant_set.drop_duplicates(inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_set.drop_duplicates(inplace=True)\n",
    "variant_set.rename(columns={\"variant_ID\": \"ID\"}, inplace=True)\n",
    "variant_set = variant_set[[\"ID\", \"esm2_mean_log_likelihood_difference\", \"esm2_perplexity_difference\", \"esm2_cosine_similarity\", \"esm2_avg_abs_difference\"]]\n",
    "\n",
    "output_file = \"../data/intermediate/esm2_wt_mut_comparison_metrics.txt\"\n",
    "os.makedirs(os.path.dirname(output_file), exist_ok=True)\n",
    "variant_set.to_csv(output_file, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>esm2_mean_log_likelihood_difference</th>\n",
       "      <th>esm2_perplexity_difference</th>\n",
       "      <th>esm2_cosine_similarity</th>\n",
       "      <th>esm2_avg_abs_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10-100057090-C-T</td>\n",
       "      <td>-0.001133</td>\n",
       "      <td>0.001560</td>\n",
       "      <td>0.999975</td>\n",
       "      <td>0.001065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10-100069757-C-T</td>\n",
       "      <td>0.000762</td>\n",
       "      <td>-0.001011</td>\n",
       "      <td>0.999961</td>\n",
       "      <td>0.001273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10-100076062-G-A</td>\n",
       "      <td>0.006985</td>\n",
       "      <td>-0.008888</td>\n",
       "      <td>0.999914</td>\n",
       "      <td>0.001779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10-100081405-G-A</td>\n",
       "      <td>-0.009512</td>\n",
       "      <td>0.012117</td>\n",
       "      <td>0.999850</td>\n",
       "      <td>0.002473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10-100152307-T-C</td>\n",
       "      <td>-0.002086</td>\n",
       "      <td>0.002513</td>\n",
       "      <td>0.999942</td>\n",
       "      <td>0.001543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408702</th>\n",
       "      <td>19-48318735-G-A</td>\n",
       "      <td>-0.022139</td>\n",
       "      <td>0.031477</td>\n",
       "      <td>0.999893</td>\n",
       "      <td>0.002803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408703</th>\n",
       "      <td>19-48318737-A-G</td>\n",
       "      <td>-0.008040</td>\n",
       "      <td>0.011297</td>\n",
       "      <td>0.999868</td>\n",
       "      <td>0.003154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408704</th>\n",
       "      <td>19-48318744-T-G</td>\n",
       "      <td>-0.002208</td>\n",
       "      <td>0.003130</td>\n",
       "      <td>0.999860</td>\n",
       "      <td>0.003399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408705</th>\n",
       "      <td>19-48318746-C-G</td>\n",
       "      <td>-0.000361</td>\n",
       "      <td>0.000516</td>\n",
       "      <td>0.999955</td>\n",
       "      <td>0.001909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408706</th>\n",
       "      <td>19-48318749-T-C</td>\n",
       "      <td>-0.011247</td>\n",
       "      <td>0.016068</td>\n",
       "      <td>0.999694</td>\n",
       "      <td>0.004806</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>408707 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      ID  esm2_mean_log_likelihood_difference  \\\n",
       "0       10-100057090-C-T                            -0.001133   \n",
       "1       10-100069757-C-T                             0.000762   \n",
       "2       10-100076062-G-A                             0.006985   \n",
       "3       10-100081405-G-A                            -0.009512   \n",
       "4       10-100152307-T-C                            -0.002086   \n",
       "...                  ...                                  ...   \n",
       "408702   19-48318735-G-A                            -0.022139   \n",
       "408703   19-48318737-A-G                            -0.008040   \n",
       "408704   19-48318744-T-G                            -0.002208   \n",
       "408705   19-48318746-C-G                            -0.000361   \n",
       "408706   19-48318749-T-C                            -0.011247   \n",
       "\n",
       "        esm2_perplexity_difference  esm2_cosine_similarity  \\\n",
       "0                         0.001560                0.999975   \n",
       "1                        -0.001011                0.999961   \n",
       "2                        -0.008888                0.999914   \n",
       "3                         0.012117                0.999850   \n",
       "4                         0.002513                0.999942   \n",
       "...                            ...                     ...   \n",
       "408702                    0.031477                0.999893   \n",
       "408703                    0.011297                0.999868   \n",
       "408704                    0.003130                0.999860   \n",
       "408705                    0.000516                0.999955   \n",
       "408706                    0.016068                0.999694   \n",
       "\n",
       "        esm2_avg_abs_difference  \n",
       "0                      0.001065  \n",
       "1                      0.001273  \n",
       "2                      0.001779  \n",
       "3                      0.002473  \n",
       "4                      0.001543  \n",
       "...                         ...  \n",
       "408702                 0.002803  \n",
       "408703                 0.003154  \n",
       "408704                 0.003399  \n",
       "408705                 0.001909  \n",
       "408706                 0.004806  \n",
       "\n",
       "[408707 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
